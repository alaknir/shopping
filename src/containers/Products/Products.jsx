import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { connect } from "react-redux";

import Product from "../../components/Product";
import ProductsHeader from "../../components/ProductsHeader";

const getMaxSizeList = products => {
  const totalSizeList = products.reduce((total, current) => {
    return total.concat(current.size);
  }, []);
  const uniqueSizeList = new Set(totalSizeList);
  return uniqueSizeList;
};

const Products = ({ products, sizeList }) => {
  const [data, setData] = useState(products);

  const onFilterChange = e => {
    const filteredData = products.filter(product =>
      product.size.includes(e.target.value)
    );
    setData(filteredData);
  };

  // eslint-disable-next-line
  useEffect(() => {
    if (!data.length) {
      setData(products);
    }
  });

  return (
    <div className="products-container">
      <ProductsHeader heading="Womens top's">
        <select
          onChange={onFilterChange}
          className="products-container__product-select"
        >
          <option value="">Filter by size</option>
          {sizeList.map(d => (
            <option value={d}>{d}</option>
          ))}
        </select>
      </ProductsHeader>
      <div className="row">
        {data.map(
          ({
            id,
            isExclusive,
            isSale,
            price,
            productImage,
            productName,
            size
          }) => (
            <Link
              key={id}
              to={`product/${id}`}
              className="col-sm-4 col-md-3 products-container__product"
            >
              <Product
                {...{
                  id,
                  isExclusive,
                  isSale,
                  price,
                  productName,
                  size
                }}
                imgSrc={`assets/products/${productImage}`}
              />
            </Link>
          )
        )}
      </div>
    </div>
  );
};

Products.propTypes = {
  products: PropTypes.shape([])
};

Products.defaultProps = {
  products: []
};

const maptStateToProps = state => {
  const products = state.products.data;
  const sizeList = Array.from(getMaxSizeList(products)) || [];
  return {
    products,
    sizeList
  };
};

export default connect(maptStateToProps)(Products);
