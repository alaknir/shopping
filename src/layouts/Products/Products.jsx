import React from "react";
import ProductsContainer from "../../containers/Products";

const Products = () => <ProductsContainer />;

export default Products;
