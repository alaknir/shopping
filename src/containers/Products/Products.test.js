import React from "react";
import { mount } from "enzyme";

import Products from "./Products";

describe("Products container", () => {
  it("Should render properly", () => {
    const component = <Products />;
    const wrapper = mount(component);
    expect(wrapper).toMatchSnapshot();
  });
});
