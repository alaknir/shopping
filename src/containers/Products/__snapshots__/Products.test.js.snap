// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products container Should render properly 1`] = `
<Products>
  <div>
    Products
  </div>
</Products>
`;
